WTrackMenu: Add menu entry for (re)analyzing tracks

fwcd opened the pull request:
In an effort to make analysis more convenient, this branch adds a new submenu to the `WTrackMenu` for analyzing tracks (previously the user had to drag-n-drop the tracks to the analyzer queue in the sidebar):

<img width="358" alt="image" src="https://user-images.githubusercontent.com/30873659/174413256-bbef4d73-a72e-4bdd-892a-6aced0136ce3.png">

Feel free to comment and make suggestions, this is my first actual code contribution to Mixxx so I am not super familiar with the codebase yet. :)

fwcd committed:
Add Reanalyze action to WTrackMenu

The reanalyze action automatically clears the beatgrid and adds the
selected tracks to the analysis queue.

ronso0 commented:
Thank you for working on this feature!

I think it makes sense to squash the two commits, why have a commit with a noop menu action.

fwcd commented:
Done!

uklotzde commented on src/widget/wtrackmenu.h:
```diff
@@ -51,8 +51,9 @@ class WTrackMenu : public QMenu {
         SearchRelated = 1 << 13,
         UpdateReplayGainFromPregain = 1 << 14,
         SelectInLibrary = 1 << 15,
+        Reanalyze = 1 << 16,
```
`Analyze` would be more general and implies re-analyzing.

These features are too detailed anyway, but that's a different topic.

Swiftb0y replied:
(Re-)Analyze?

uklotzde commented on src/widget/wtrackmenu.cpp:
```diff
@@ -2140,6 +2168,12 @@ bool WTrackMenu::featureIsEnabled(Feature flag) const {
     case Feature::Metadata:
         return m_pTrackModel->hasCapabilities(TrackModel::Capability::EditMetadata);
+    case Feature::Reanalyze:
+        // TODO: Do we actually need the EditMetadata capability here?
```
When analyzing tracks their metadata needs to be updated for storing the results. Clearing the beat grid beforehand doesn't really matter imho.

`TrackModel::Capability::Analyze` should already imply `TrackModel::Capability::EditMetadata`. These capability flags are inconsistent and overlapping in their meaning. Everyone just extended it to keep it working. Unrelated, has to be resolved separately.

ronso0 replied:
> Clearing the beat grid beforehand doesn't really matter imho.

Yes, it does. Otherwise the beats where not reanalyzed inmy tests.

uklotzde reviewed (commented):
Thanks for starting to on work on this feature!

Adding a new top-level entry to the already crowded context menu would not be my first choice. Ideas how to restructure the menu are welcome. I don't have any.

ronso0 commented:
> the action automatically clears bpm/beatgrid and adds the selected tracks to the analysis queue. The idea behind including the 'clear beatgrid' action is that (at least in my experience) this is a much more common operation than simply analyzing tracks that have no grid yet.

I think there are different use cases for (re-)analyzing which require different reset actions:
* new tracks and don't like dragging to Analysis? simply analyze
* edited the beat grid and messed it up? clear beats
* track file was edited? clear the waveform, too
* ?

Could the different use cases be served with dialog incl. checkboxes for required reset actions?

> Ideas how to restructure the menu are welcome. I don't have any.

I agree Analyze is not really "urgent" enough to be a top-level action.
What about renaming "Reset" > "Reset / Analyze" and have all actions there?

Could anything go wrong when trying to reset beats? Beats could be locked for some of many selected tracks.. then re-analysis would be pointless (the result irritating/unexpected) without UI feedback.
Check `isAnyTrackBpmLocked()`.
Probably better to process tracks (clearX, queue to be analyzed) one by one, like `for (trackId : trackIds) {resetBeats; addToAnalysis;}`, like in `ClearAllPerformanceMetadataTrackPointerOperation`?

fwcd commented:
While I agree that the menu is pretty crowded, burying analysis in a submenu IMO negates the convenience that this patch is intended to bring to a certain degree.

I'd actually argue that Analysis would make sense as a new top-level menu, given that analysis is a pretty prominent action a user takes on his tracks (and rather underrepresented in the UI currently). Additionally, there are several 'flavors' of analysis that a user might want to perform on-the-fly, that could fit into a top-level analysis menu:

- 'Just' analyzing
- Resetting beats, then analyzing
- Analyzing with constant bpm (overriding the setting)
- Analyzing with variable bpm (also overriding the setting)

The latter two points would be out-of-scope for this PR, but planning ahead would make sense IMO.

ronso0 commented:
> 'Just' analyzing
    Resetting beats, then analyzing
    Analyzing with constant bpm (overriding the setting)
    Analyzing with variable bpm (also overriding the setting)

The first two already justify having an Analyze menu (not Analyze action + Analyze menu) instead of the dialog + checkbox I proposed.

daschuer commented:
Is re-analyzing because of messed up beat grid the most common use case? 

If not I can think of manual reset the desired properties first. On the other hand all other features are pretty much read only and can be re-analyzed without a risk. 

The use case of reanalyze after editing the track sound itself is special, because it is probably desired to move the waveform along with cue points and beat-grid. This is not a core topic for this PR.

So for now I vote for this solution:
- Analyze (missing properties)
- Re-Analyze Const BPM
- Re-Analyze Variable BPM

The next step would be to verify the offset, before Re-Analyze a new beat-grid, else the waveform and cue point will not match the nevw beatgrid. 

Later we may add:
- Re-Analyze Offset/Edits

fwcd commented:
I would like to keep per-track constant/variable analysis out of this PR, since that seems to be quite a bit more complex to implement. Also I wouldn't do anything super fancy either in this PR, to keep the scope manageable, perhaps just an Analysis menu with two entries

- Analyze
- Reanalyze (with current settings)

And everything else could be determined in the future?

daschuer commented:
That sounds reasonable

In this case it would be:
* Analyze 
* Re-Analyze BPM

fwcd committed:
Rename feature to Analyze

fwcd committed:
Make analysis a submenu of WTrackMenu

fwcd committed:
Add Analyze capability to PlaylistTableModel

fwcd committed:
Remove Analyze capability from BrowseTableModel

...for consistency, since it doesn't include the EditMetadata capability
either.

fwcd commented:
I have updated the menu entry to be a submenu containing both the 'simple' analysis action and one that clears the beats beforehand:

<img width="358" alt="image" src="https://user-images.githubusercontent.com/30873659/174413194-5cd8177b-e95c-4fb8-b85f-4546c2a14efb.png">

daschuer commented:
This not just adds the tracks to the queue, it also starts the analyze, which is the main part of the use case. 
How will this look like after the next step that allows to select const/non const? 

What about this: 

* Analyze 
** Missing properties
** BPM


Than we can change it later to: 
* BPM assume const
* BPM variable

fwcd commented:
I think calling it 'BPM' would be a bit misleading since it would analyze all other (missing) properties too. The mental model behind 'Add to Queue' is that the analyzer is already running (similar to dragging tracks into the 'analyzer' in the sidebar which to me is a queueing operation). Since that might not match the technical side of things, I agree that we probably want a better name.

daschuer commented:
Maybe my proposed alternative is not the best. How will your proposal look like if we add the const / non const thing? 

Your proposal would be translated like this: 
* Analysieren > Zur Warteschlange hinzufügen
* Analysieren > bpm zurücksetzen und zur Warteschlange hinzufügen

IMHO the queue thing sounds like exposing an implementation detail to me.  

I can also think of moving the BPM part to the Adjust BPM menu  
* Analyze 
* Adjust BPM -> Re-Analyze

fwcd commented:
Hm, I agree that the German translation is quite verbose, but would still prefer to keep analysis-related things in a single menu. What do you think about:

- Analyze (top-level menu)
  - Analyze
  - Reanalyze

adding

- Reanalyze (constant BPM)
- Reanalyze (variable BPM)

or similar in the future?

daschuer commented:
That works for me as well. 

I would keep the key out of this game for now. 
It might interesting once we have changing key support. Can be added later.

fwcd committed:
Update menu action names to 'Analyze' and 'Reanalyze'

daschuer reviewed (approved):
LGTM, Thank you. 
@uklotzde @ronso0: merge?

uklotzde reviewed (approved):
@daschuer Didn't check again, just merge. What I have seen so far from @fwcd all looked solid ;)

daschuer commented:
Tank you.
